### PR TITLE
gh-102120: Added an iter function that doesn't cache

### PR DESCRIPTION
**Pull request by PurityLake**
Refers to #102120

This is a proposed solution to the linked issue which is to provide a way to iterate over a tar object so that it doesn't cache.

My suggestion is to reuse the code from `Tarfile._next()` to yield a tarinfo each time instead of caching it. This way it doesn't take up a tonne of memory.

I am looking for feedback on this so I will mark this as a draft.

Edit:

As from suggestions a parameter called stream can be passed to the object which will make sure nothing is cached. I will add a code example later today.

* Issue: gh-102120

### CODE REVIEW

**Comment by ethanfurman** [comment]
I would rather add a new keyword to the `TarFile` signature -- perhaps `stream` -- and update the relevent locations in the code to not update `members`.

**Comment by PurityLake** [comment]
That would probably be better

**Comment by PurityLake** [comment]
I'll be adding some tests later today

**Comment by PurityLake** [comment]
I believe this is working as intended now. I'm open to criticism in the implementation

File: Doc/library/tarfile.rst [hunk 1]
```diff
@@ -378,6 +381,8 @@ be finalized; only the internally used file object will be closed. See the
    .. versionchanged:: 3.6
       The *name* parameter accepts a :term:`path-like object`.
 
+   .. versionchanged:: 3.12
+      The *stream* parameter has been added.
```
Comment (ethanfurman) [inline Doc/library/tarfile.rst]:
```suggestion
   .. versionchanged:: 3.13
      Add the *stream* parameter.
```

**Review by ethanfurman — CHANGES_REQUESTED** [review]
Overall, looks good.  I missed feature-freeze, though, so this will have to go in 3.13.  Update the `3.12`s to `3.13`s and I'll get it merged.

**Comment by PurityLake** [comment]
I have made the requested changes; please review again

**Review by ethanfurman — APPROVED** [review]

**Pull request merged**